### PR TITLE
feat(plg-onboard): waitlist authenticated (login/SSO) sites during onboarding

### PR DESCRIPTION
**Pull request by Kanishkavijay39**
## 1. Abstract
Detects authenticated (login/SSO-gated) sites during PLG onboarding and rejects them outright — before any site/entitlement is provisioned — instead of auto-onboarding a site ASO cannot audit.

## 2. Reasoning
During PLG onboarding, login-gated sites were auto-onboarded into ASO even though ASO cannot support them: OpTel/RUM data flows (so a CWV opportunity can appear), but crawl-based audits (broken-backlinks, alt-text, scrape) hit the auth wall and return the login page. The site lands partially functional, producing an incomplete optimizer experience and requiring manual cleanup. The reported case was TATA Steel's `sampoornanxt.tatasteel.com`, whose front door redirects an anonymous crawler to a login page. Onboarding had no notion of an auth wall, so nothing stopped it.

## 3. High-level overview of the changes
Adds an auth-wall detection gate to the PLG onboarding flow that rejects login/SSO-gated domains outright.

- A new probe fetches the site's front door anonymously during the onboarding crawl phase, immediately after the bot-blocker check and **before** any Site or entitlement is provisioned. Redirects are followed **manually** and every hop's host is validated with `isSafeDomain()` before it is requested, so an attacker-controlled 3xx on a user-supplied domain cannot steer the probe into an internal/private host (SSRF-safe). The resolved URL that goes into the reason is further reduced by `sanitizeUrlForReason()` (scheme+host+path only, query/fragment and mrkdwn/control chars stripped, length-capped).
- Detection is deliberately high-precision to avoid ever blocking a legitimate public customer. A site is flagged only on a strong signal: an HTTP 401, a front door that resolves to a login/SSO URL, a known identity-provider host (Okta / Auth0 / Microsoft / Google / Adobe IMS / OneLogin / Ping / Cloudflare Access), or a landing page that is unambiguously a login form (a password field together with a login/sign-in page title). A public homepage with an incidental account/login widget is not flagged. If the probe errors or times out it fails open (onboarding proceeds), so a flaky probe never blocks a real customer.
- Behaviour delta: previously an auth-gated domain was auto-onboarded (Site + ASO entitlement + audits) → now it is set to terminal `status: REJECTED` with a clear `waitlistReason` explaining it requires authentication, and no site/entitlement is created. The PLG onboarding Slack notification renders the reason for auto-rejections, so the team sees it in the `plg_onboardings` channel.
- Why reject rather than waitlist: unlike every other waitlist reason (bot-blocker → allowlist IPs, non-prod domain → give the prod domain, AEM-site-check → confirm delivery type, domain-already-assigned → move/alternate), an auth wall has **no self-service remediation** — ASO's crawlers fundamentally cannot get behind a login. Holding these for manual review would only ever produce a rubber-stamp reject, so they are rejected at the gate. Trade-off: a false-positive detection is a hard reject with no in-product override; mitigated by the high-precision + fail-open + SSRF-guarded probe. A retry of `POST /plg/onboard` re-probes, so a site that later becomes publicly crawlable can still onboard.
- Not in this PR (deferred, was the optional acceptance criterion): backfilling/offboarding sites already onboarded behind auth (e.g. the existing TATA Steel site) — that needs its own enumerate-and-re-probe approach. User-facing onboarding-UI / back-office copy lives in other repos; this PR provides the machine-readable `status: REJECTED` + `waitlistReason` those surfaces render.

## 4. Required information
- Jira / issue: [SITES-47346](https://jira.corp.adobe.com/browse/SITES-47346)

## 7. Additional information outside the code
Confirmed the detection signal against the live reported site (read-only, anonymous GET):
- `curl -sIL https://sampoornanxt.tatasteel.com/` follows one redirect and resolves to `https://sampoornanxt.tatasteel.com/sampoorna/login.html` (HTTP 200).
- The resolved landing page's HTML contains `<title>login</title>`, an `<input type="password">`, and "sign in" content — i.e. both the login-URL signal and the login-form signal the probe keys on. This is the case that motivated the ticket and it is now flagged.

## 8. Test plan
(a) Verified beyond automated tests: probed the live TATA Steel domain (above) to ground the detection heuristic in the real login-wall shape, and exercised the onboarding flow behaviourally — an auth-detected domain is set to `REJECTED` with the authentication reason and no Site/entitlement is created; the probe fails open on a network/timeout error; a normal public domain still reaches `ONBOARDED`.

(b) Per-environment verification (dev → stage → prod):
- Pick a known login-gated domain (one that redirects anonymous requests to a login/SSO page).
- `POST /plg/onboard` for it (see the `plg-onboard` skill for the admin-token invocation). Expect `200` with `status: "REJECTED"` and a `waitlistReason` containing "requires authentication (login/SSO)". Confirm no Site/entitlement was provisioned.
- Confirm the `plg_onboardings` Slack channel shows the Rejected notification with the reason line.
- Regression: onboard a normal public production domain and confirm it still reaches `ONBOARDED` (the probe does not false-positive).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Change Management

```yaml
cm-assessment: v1
changeType: standard
impact: unnoticeable
risk: minor
changeApprovedBy: ["Anshul Kumar", "Dominique Jäggi"]
rationale: "Adds a fail-open, SSRF-guarded detection gate for customer sites requiring login; rejects with a reversible retry path, no change to SpaceCat's own auth/access-control boundary."
```

### CODE REVIEW

File: src/support/detect-auth-wall.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+/**
+ * Fetches the base URL anonymously (following redirects) and classifies whether its front
+ * door is a login / SSO wall.
+ *
+ * @param {{ baseUrl: string, log?: object }} params
+ * @param {{ fetch?: Function }} [deps] - injectable fetch for testing; defaults to global fetch.
+ * @returns {Promise<{authenticated: boolean, signal: string|null, finalUrl: string|null,
+ *   status: number|null}>}
```
Comment (MysticatBot) [inline src/support/detect-auth-wall.js:158]: issue (blocking): **SSRF via open redirect - redirect target not validated.**

The probe uses `redirect: 'follow'` but only the initial domain is validated by `isSafeDomain()` upstream. A public domain under attacker control can 3xx-redirect to internal hosts (VPC services, metadata endpoints). The `response.url` (final URL after redirects) is then stored in `waitlistReason` and exposed via `PlgOnboardingDto.toJSON()`, creating an information-disclosure oracle for authenticated users.

**Why it matters:** An authenticated attacker can probe internal infrastructure reachability and learn internal redirect targets via the public waitlist reason field.

**How to fix:** Validate the resolved hostname after fetch completes, before processing the response:

```javascript
const resolvedHost = new URL(response.url).hostname;
if (net.isIP(resolvedHost) || !isSafeDomain(resolvedHost)) {
  log?.warn?.(`Auth-wall probe redirected to disallowed host: ${response.url}`);
  return notAuthenticated();
}
```

Alternatively, redact `finalUrl` from the user-facing `waitlistReason` (log it server-side for admin diagnosis only).

File: src/controllers/plg/plg-onboarding/notifications.js [hunk 2]
```diff
@@ -44,6 +46,9 @@ export function deriveCheckKey(onboarding) {
   if (waitlistReason.includes(NON_PROD_DOMAIN)) {
     return REVIEW_REASONS.NON_PROD_DOMAIN;
   }
+  if (waitlistReason.includes('requires authentication')) {
```
Comment (MysticatBot) [inline src/controllers/plg/plg-onboarding/notifications.js]: issue (blocking): **`deriveCheckKey` uses a hardcoded substring instead of the exported constant.**

```javascript
if (waitlistReason.includes('requires authentication')) {
```

The `AUTHENTICATED_SITE` constant is defined and exported in the same file. The other branches use their respective constants (`DOMAIN_ALREADY_ONBOARDED_IN_ORG`, `NON_PROD_DOMAIN`). Using a hardcoded substring is fragile: if the `AUTHENTICATED_SITE` wording changes, `deriveCheckKey` would silently return `null` and the admin bypass handler would never be routed.

**How to fix:** Use the constant:
```javascript
if (waitlistReason.includes(AUTHENTICATED_SITE)) {
  return REVIEW_REASONS.AUTHENTICATED_SITE;
}
```
Or extract a stable machine-readable prefix that cannot drift from the assembly site.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @Kanishkavijay39,

**Verdict: Request changes** - two actionable issues around redirect-following safety and string-match routing fragility.
**Complexity: HIGH** - large diff; API surface + security signal (outbound fetch on user-supplied domain).
_Changes: Adds auth-wall detection to PLG onboarding that waitlists login/SSO-gated sites before provisioning, with admin bypass support (15 files)._
Note: CI checks are currently failing - resolve before merge.

## Must fix before merge
1. [Important] SSRF via open redirect - redirect target hostname not validated after fetch follows redirects - `src/support/detect-auth-wall.js:89` (details inline)
2. [Important] `deriveCheckKey` uses hardcoded substring instead of the co-located constant - `src/controllers/plg/plg-onboarding/notifications.js:49` (details inline)

<details>
<summary>Non-blocking (5): minor issues and suggestions</summary>

- nit: `LOGIN_URL_PATTERN` matches bare `/auth` segment which could false-positive on `/auth/callback` or `/auth/verify-email` paths - consider tightening or adding a test documenting accepted behavior - `src/support/detect-auth-wall.js:28`
- nit: `steps.authWallChecked = true` is set but never read downstream - clarify if informational or remove - `src/controllers/plg/plg-onboarding/onboarding-flow.js:675`
- suggestion: fuse or run concurrently the bot-blocker and auth-wall probes (both fetch the same baseURL with redirect-follow) to avoid stacking two 15s timeouts on the critical onboarding path - `src/controllers/plg/plg-onboarding/onboarding-flow.js:660`
- suggestion: reduce `BODY_SCAN_LIMIT` from 500KB to 100-200KB - login forms appear in early markup; a lower limit reduces memory pressure without losing detection - `src/support/detect-auth-wall.js:15`
- suggestion: add a test documenting that HTTP 403 alone is NOT flagged (common auth response, currently falls through to body scan) to prevent accidental regressions - `test/support/detect-auth-wall.test.js`

</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 3m 7s | <b>Cost:</b> $8.00 | <b>Commit:</b> <code>f2907e725226146b996326988a196c2cf03313f1</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/support/detect-auth-wall.js (hunk 1 above)
Comment (Kanishkavijay39) [reply src/support/detect-auth-wall.js:158]: Fixed in a6a5493. Chose the stronger option: the probe now follows redirects **manually** (`redirect: 'manual'`) and validates every hop's host with `isSafeDomain()` **before** requesting it, so an attacker-controlled 3xx can't make the probe reach an internal host at all (not just post-hoc detection). A consequence is that `finalUrl` is now guaranteed to be a public host, so keeping it in `waitlistReason` discloses nothing internal. Added SSRF-guard tests: redirect to cloud-metadata/RFC1918 hosts, a private base URL, and an over-long redirect chain — all fail open to 'not authenticated' without fetching or surfacing the disallowed target.

File: src/controllers/plg/plg-onboarding/notifications.js (hunk 2 above)
Comment (Kanishkavijay39) [reply src/controllers/plg/plg-onboarding/notifications.js]: Fixed in a6a5493 — `deriveCheckKey` now matches on the exported `AUTHENTICATED_SITE` constant instead of the hardcoded `'requires authentication'` substring, so the wording can't drift out from under the admin-bypass routing.

**Comment by Kanishkavijay39** [comment]
### Review addressed (a6a5493)

**Blocking**
- **SSRF via open redirect** — probe now follows redirects manually and validates every hop's host with `isSafeDomain()` before requesting it; the internal host is never reached. `finalUrl` is consequently always public, so it stays in the waitlist reason safely. New SSRF-guard tests added.
- **`deriveCheckKey` hardcoded substring** — now uses the exported `AUTHENTICATED_SITE` constant.

**Non-blocking**
- `BODY_SCAN_LIMIT` 500KB → 200KB.
- Added a test documenting that a bare **HTTP 403** is not flagged (falls through to body scan).
- Added a test documenting the accepted behavior that an `/auth/*` front door **is** treated as a login endpoint (kept the pattern; tightening risks missing real login redirects).
- `steps.authWallChecked` — documented as an intentional informational audit breadcrumb (mirrors the other `steps.*` flags).
- **Deferred:** fusing the bot-blocker + auth-wall probes to run concurrently. The bot-blocker check gates an early return before the auth-wall probe, so naive concurrency would change behavior and fire an extra request on the blocked path — not worth the complexity in this change.

File: src/controllers/plg/plg-onboarding/onboarding-flow.js [hunk 3]
```diff
@@ -655,6 +657,27 @@ export async function performAsoPlgOnboarding({
...
     }
 
+    // Step 4b: Authenticated-site check — ASO cannot audit login/SSO-gated sites, so route
+    // them to the manual-review waitlist before any site/entitlement is provisioned.
+    if (!steps.authWallCheckBypassed) {
+      const authWall = await detectAuthWall({ baseUrl: baseURL, log });
+      if (authWall.authenticated) {
+        log.info(`Domain ${domain} appears to require authentication (signal: ${authWall.signal}), moving to waitlist`);
+        let waitlistReason = `Domain ${domain} ${AUTHENTICATED_SITE} (detected: ${authWall.signal}`;
+        waitlistReason += authWall.finalUrl ? `, resolved to ${authWall.finalUrl}).` : ').';
```
Comment (MysticatBot) [inline src/controllers/plg/plg-onboarding/onboarding-flow.js]: issue (blocking): `finalUrl` is attacker-controlled (it is the final URL after redirect-following against a user-supplied domain). The path, query string, and fragment are entirely under the attacker's control.

This value is interpolated raw into `waitlistReason`:
```js
let waitlistReason = `Domain ${domain} ${AUTHENTICATED_SITE} (detected: ${authWall.signal}`;
waitlistReason += authWall.finalUrl ? `, resolved to ${authWall.finalUrl}).` : ')';
```

`waitlistReason` is then:
1. Persisted to the database via `onboarding.setWaitlistReason()`
2. Returned in the onboarding DTO (visible to API consumers)
3. Posted to Slack via `persistAndNotify` - Slack renders mrkdwn, so crafted path/query can inject formatting, links, or @-mentions

Fix: sanitize `finalUrl` before interpolation - strip query and fragment (`new URL(finalUrl).origin + new URL(finalUrl).pathname`), and cap length (e.g. 200 chars) to prevent log/DB bloat from long paths.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @Kanishkavijay39,

> :warning: Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - one remaining issue around unsanitized attacker-controlled URL data in the waitlist reason string.
**Complexity: HIGH** - large diff; API surface + security signal (outbound fetch on user-supplied domain).
_Changes: Adds auth-wall detection to PLG onboarding with SSRF-safe redirect following, waitlists login/SSO-gated sites, and adds admin bypass support (15 files)._

## Must fix before merge
1. [Important] `finalUrl` interpolated into `waitlistReason` without sanitization - attacker-controlled path/query/fragment persisted to DB and forwarded to Slack (mrkdwn injection) - `src/controllers/plg/plg-onboarding/onboarding-flow.js:667` (details inline)

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: Layer inversion - `src/support/detect-auth-wall.js` imports `isSafeDomain` from `src/controllers/`. Consider moving the shared utility to `src/support/` or `src/utils/` - `src/support/detect-auth-wall.js:14`
- nit: First `IDP_HOST_PATTERNS` entry (`login.microsoftonline.com`) is redundant - the second pattern (`login.microsoft(?:online)?.com`) already covers it - `src/support/detect-auth-wall.js:46`
- nit: `LOGIN_URL_PATTERN` matches bare `auth` path segment, which flags `/auth/callback` paths. Tests document this as accepted behavior, but worth a docstring note if false-positive rates become material - `src/support/detect-auth-wall.js:28`

</details>

## Previously flagged, now resolved
- SSRF via open redirect - fixed: `fetchGuardingRedirects` now validates each hop's hostname with `isSafeDomain()` before issuing the request
- `deriveCheckKey` hardcoded substring - fixed: now uses the co-located `AUTHENTICATED_SITE` constant
- `LOGIN_URL_PATTERN` bare `/auth` match - documented as accepted behavior with explicit test coverage
- `steps.authWallChecked` audit-trail breadcrumb - retained as informational, consistent with other `steps.*` flags
- Bot-blocker/auth-wall probe fusion - addressed as out-of-scope for this PR (separate HTTP clients with different redirect semantics)
- `BODY_SCAN_LIMIT` sizing - retained at 500KB as conservative default, consistent with existing patterns
- 403 non-flag test - added implicit coverage through the detector's explicit status-code handling

File: src/controllers/plg/plg-onboarding/onboarding-flow.js (hunk 3 above)
Comment (Kanishkavijay39) [reply src/controllers/plg/plg-onboarding/onboarding-flow.js]: resolved

Comment (Kanishkavijay39) [reply src/controllers/plg/plg-onboarding/onboarding-flow.js]: Fixed in 3bcee1d. `authWall.finalUrl` is now reduced via a new `sanitizeUrlForReason()` before it enters `waitlistReason`: it keeps only `scheme://host/path` (the host is already SSRF-validated public), **drops the attacker-controlled query + fragment entirely**, strips Slack-mrkdwn/control characters (`< > | ` * _ ~` and CR/LF/TAB — note the URL parser already percent-encodes `<`, `>` and backtick in the path), and caps length at 256. So the value persisted to the DB and sent to Slack can't carry a link/mention injection or break out of the reason string. Unit-tested in `test/support/url-safety.test.js`.

**Comment by Kanishkavijay39** [comment]
### Second review addressed (3bcee1d)

**Blocking**
- **Unsanitized `finalUrl` in `waitlistReason`** — added `sanitizeUrlForReason()`: keeps only `scheme://host/path`, drops the attacker-controlled query + fragment, strips Slack-mrkdwn/control chars, caps at 256. Applied before persist + Slack. Unit-tested.

**Non-blocking**
- **Layer inversion** — moved `isSafeDomain` into `src/support/url-safety.js` (shared, correct layer) and re-export it from `plg-onboarding/validation.js`, so `detect-auth-wall.js` no longer imports up into a controller. `sanitizeUrlForReason` lives there too.
- **Redundant IDP pattern** — dropped `login.microsoftonline.com` (already covered by `login.microsoft(online)?.com`).
- **`/auth` bare segment** — added a docstring note documenting it as accepted front-door behavior (test already covers it).

Verification: `url-safety` + `detect-auth-wall` + `validation` suites green (92), full PLG onboarding suite green (432), lint + strict type-check clean.

**Review by MysticatBot — APPROVED** [review]
Hey @Kanishkavijay39,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - all prior blocking findings are addressed; remaining items are non-blocking.
**Complexity: HIGH** - large diff; API surface + security signal (outbound fetch on user-supplied domain).
_Changes: Sanitizes attacker-controlled finalUrl before persistence/Slack, moves isSafeDomain to shared support module to fix layer inversion, and removes a redundant IDP host pattern (18 files)._

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: isSafeDomain JSDoc contract ("callers MUST invoke prepareDomain() and isValidDomain() BEFORE this function") is stale for its new caller in detect-auth-wall.js, which passes a bare hostname from new URL().hostname instead - `src/support/url-safety.js:22`
- suggestion: add integration tests in test/it/ for the authenticated-site waitlist flow per the repo's CLAUDE.md convention for modified endpoints - `test/it/`
- suggestion: add a test that imports isSafeDomain from the validation.js re-export path to prevent silent breakage if the re-export is accidentally removed - `test/controllers/plg/plg-onboarding/validation.test.js`

</details>

## Previously flagged, now resolved
- finalUrl interpolated into waitlistReason without sanitization - fixed: sanitizeUrlForReason drops query/fragment, strips Slack mrkdwn chars, and caps length
- Layer inversion (detect-auth-wall importing from controller) - fixed: isSafeDomain moved to src/support/url-safety.js
- Redundant login.microsoftonline.com IDP pattern - fixed: removed (covered by login.microsoft(?:online)?.com)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 58s | <b>Cost:</b> $6.90 | <b>Commit:</b> <code>3bcee1d9abd33165fc9e86132b29e5230c19d6fe</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/controllers/plg/plg-onboarding/onboarding-flow.js [hunk 4]
```diff
@@ -655,6 +658,30 @@ export async function performAsoPlgOnboarding({
...
+        onboarding.setWaitlistReason(waitlistReason);
+        onboarding.setSiteId(site?.getId() || null);
+        onboarding.setSteps(steps);
+        await persistAndNotify(onboarding, context);
+        return onboarding;
+      }
+      // Informational audit-trail breadcrumb (persisted on the onboarding record like the
+      // other `steps.*` flags): records that the auth-wall probe ran and the front door was
+      // public. Not read back in the flow; kept for post-hoc diagnosis of onboarding runs.
+      steps.authWallChecked = true;
```
Comment (anshulk-public) [inline src/controllers/plg/plg-onboarding/onboarding-flow.js]: Please add steps.authWallChecked and steps.authWallChecked in schema if not added already. Apart from that it looks good to me.

**Review by anshulk-public — APPROVED** [review]
LGTM

**Review by MysticatBot — DISMISSED** [review]
@/tmp/mysticat-workspace-sz_2_55j/review-body-3142.txt

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @Kanishkavijay39,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - PR description claims admin UPHELD/BYPASSED support that the code does not implement.
**Complexity: HIGH** - large diff; API surface + security signal (outbound fetch on user-supplied domain).
_Changes: Adds auth-wall detection to PLG onboarding that rejects login/SSO-gated sites before provisioning, with SSRF-safe redirect following and URL sanitization (18 files)._
Note: CI checks are currently pending - verify before merge.

## Must fix before merge
1. [Important] PR description claims admin BYPASSED/UPHELD support for auth-detected sites, but the code sets REJECTED (terminal, no admin action possible) - `src/controllers/plg/plg-onboarding/onboarding-flow.js:701` (details inline)

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: `isSafeDomain` JSDoc (url-safety.js:22) states "callers MUST invoke prepareDomain() and isValidDomain() BEFORE this function" but the new caller in detect-auth-wall.js passes a bare hostname from `new URL().hostname` without either prerequisite. The call is safe (URL parser strips scheme/path), but the JSDoc misleads future callers. - `src/support/url-safety.js:22`
- nit: `isHtmlContentType` treats null/undefined content-type as HTML (falls through to body scan). Behavior is reasonable but has no explicit test - consider adding a test with `contentType: null` and a login-form body. - `src/support/detect-auth-wall.js:95`
- suggestion: Assert that `steps.authWallChecked = true` is set on the happy path (public site proceeds to provisioning) to document the audit-trail breadcrumb behavior. - `test/controllers/plg/plg-onboarding/onboarding-flow-core.test.js`
- nit: PR body section 2 says "set to WAITLISTED" but the code sets REJECTED. Section 3 and the SKILL.md correctly say "reject outright". Update sections 2 and 3 of the PR body to match the implemented behavior.

</details>

## Previously flagged, now resolved
- SSRF via open redirect - fixed: fetchGuardingRedirects validates each hop's hostname with isSafeDomain() before requesting it
- deriveCheckKey hardcoded substring - fixed: uses the AUTHENTICATED_SITE constant
- Unsanitized finalUrl in waitlistReason - fixed: sanitizeUrlForReason strips query/fragment, mrkdwn chars, caps at 256
- Layer inversion (detect-auth-wall importing from controller) - fixed: isSafeDomain moved to src/support/url-safety.js
- Redundant IDP host pattern - fixed: removed (covered by existing regex)


---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 18s | <b>Cost:</b> $12.22 | <b>Commit:</b> <code>1d77c474c5fb86fede9463707a28250ea5d94484</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/controllers/plg/plg-onboarding/onboarding-flow.js [hunk 5]
```diff
@@ -683,6 +686,30 @@ export async function performAsoPlgOnboarding({
...
+    // provisioned.
+    const authWall = await detectAuthWall({ baseUrl: baseURL, log });
+    if (authWall.authenticated) {
+      log.info(`Domain ${domain} appears to require authentication (signal: ${authWall.signal}), rejecting`);
+      // finalUrl is host-validated (public) but its path/query/fragment are caller-controlled;
+      // reduce it before it is persisted and forwarded to Slack (mrkdwn) to avoid injection.
+      const safeFinalUrl = authWall.finalUrl ? sanitizeUrlForReason(authWall.finalUrl) : '';
+      let rejectionReason = `Domain ${domain} ${AUTHENTICATED_SITE} (detected: ${authWall.signal}`;
+      rejectionReason += safeFinalUrl ? `, resolved to ${safeFinalUrl}).` : ').';
+      onboarding.setStatus(STATUSES.REJECTED);
```
Comment (MysticatBot) [inline src/controllers/plg/plg-onboarding/onboarding-flow.js:701]: issue (blocking): **PR description claims admin BYPASSED/UPHELD support that does not exist in the code.**

The code sets `STATUSES.REJECTED` (line 701), not `STATUSES.WAITLISTED`. The `AUTHENTICATED_SITE` reason string is deliberately defined outside `REVIEW_REASONS` (the comment at notifications.js:42 says "a terminal-reason string, not a REVIEW_REASONS key"). The admin PATCH handler guards with `if (status !== STATUSES.WAITLISTED)` and returns `badRequest`, so both UPHELD and BYPASSED decisions are blocked for auth-detected records.

But the PR description section 3 says: "UPHELD rejects it; a new BYPASSED handler force-onboards it by re-running the flow with the auth-wall check skipped." The test plan (section 8b) also instructs reviewers to verify the BYPASSED and UPHELD paths. Neither path works.

**Why it matters:** Operators trusting the PR description will expect an admin bypass path that does not exist. The code's design (terminal rejection, no bypass) is internally coherent and arguably correct, but the PR description contradicts it.

**How to fix (pick one):**
- If auth-site rejections are truly terminal (no remediation): update PR body sections 2, 3, and 8b to match the code. Remove the BYPASSED/UPHELD claims.
- If admin bypass is intended: add `AUTHENTICATED_SITE` to `REVIEW_REASONS`, add a `deriveCheckKey` branch for it, and add a `bypassAuthenticatedSite` case to the admin switch block (using the existing `steps.authWallCheckBypassed` guard to skip the probe on re-run).

Comment (Kanishkavijay39) [reply src/controllers/plg/plg-onboarding/onboarding-flow.js:701]: updated

**Review by MysticatBot — APPROVED** [review]
@/tmp/mysticat-workspace-1mw78_ms/review-body-3142.txt

**Review by MysticatBot — DISMISSED** [review]
**Mysticat could not complete a verified review of this pull request.**

This is a service-side gap, not a finding about your code: no binding APPROVE or REQUEST_CHANGES review could be produced and verified for the current commit.

- Current head: `67140ad21ce100b6b8450fbcab3ee576a68bef84`
- Failure category: `completion_review_mismatch`
- Request ID: `fc9a81b0-aab6-11f1-9bf0-f4466c874532`
- Cost: $7.66
- Duration: 2m 34s

Please re-request a review (e.g. push a new commit or re-request from the configured reviewer) to try again.
